src/Spice86.Core/Emulator/OperatingSystem/DosMemoryManager.cs

```diff
@@ -162,10 +182,235 @@ public DosErrorCode TryModifyBlock(in ushort blockSegment, in ushort requestedSi
+                    block.Size == size.MinSizeInParagraphs ? "required" : "requested",
+                    block.Size,
+                    block.AllocationSizeInBytes,
+                    ConvertUtils.ToHex16(block.DataBlockSegment));
+            }
+        } else if (_loggerService.IsEnabled(LogEventLevel.Error)) {
+            _loggerService.Error(
+                "{SizeInParagraphs} paragraphs ({SizeInBytes} bytes) are not available at {PspSegment} to load program",
+                size.MinSizeInParagraphs,
+                size.MinSizeInParagraphs * 16,
+                ConvertUtils.ToHex16(pspSegment));
+        }
+
+        return block;
+    }
+
+    /// <summary>
+    /// Range the specifies a minimum and maximum size to allocate for a block of memory.
+    /// </summary>
+    /// <remarks>
+    /// This is a helper type for the memory manager that is designed to support allocating a block
+    /// at is at most the maximum size specified in this struct and at least the minimum size
+    /// specified in this struct. It is primarily intended to support the min/max alloc concept in
+    /// EXE files to allow the calculated values to be easily passed around together and to make it
+    /// easier to identify where min/max allocation is requested inernally in the code.
+    /// </remarks>
+    private struct AllocRange {
+        /// <summary>
+        /// Minimum number of paragraphs that <em>must</em> be allocated.
+        /// </summary>
+        public ushort MinSizeInParagraphs { get; set; }
+
+        /// <summary>
+        /// Maximum number of paragraphs that <em>may</em> be allocated if they are available.
+        /// </summary>
+        public ushort MaxSizeInParagraphs { get; set; }
+    }
+
+    /// <summary>
+    /// Calculates the minimum size required to load an EXE and the maximum requested size that it
+    /// would like if it's available.
+    /// </summary>
+    /// <remarks>
+    /// This function <em>does not</em> allocate any memory. It just calculate the required sizes
+    /// that will need to be allocated. It's a member of this class rather than being calculated in
+    /// the <see cref="DosExeFile"/> class because it requires insight into the current free memory
+    /// blocks in some cases. It's more complicated that it may seem on the surface.
+    /// </remarks>
+    /// <param name="exeFile">EXE file header that defines the amount of space we need.</param>
+    /// <param name="pspSegment">Segment address where the PSP before the EXE will be loaded.</param>
+    /// <returns>The calculated minimum required and maximum requested allocation sizes.</returns>
+    private AllocRange CalculateSizeForExe(DosExeFile exeFile, ushort pspSegment) {
+        // Every program requires at least enough space for itself and the 16 paragraph (256 byte)
+        // PSP that precedes it.
+        ushort baseSizeInParagraphs = (ushort)(exeFile.ProgramSizeInParagraphs + 0x10);
+
+        var size = new AllocRange {
```

Comment:
I would make the datastructure read only and instanciate it at the end of the function at return time.

Code that computes new max size based on freeSizeInParagraphs could just be a ternary

```return new AllocRange(
    (ushort)(baseSizeInParagraphs + exeFile.MinAlloc),
    freeSizeInParagraphs==0?
        (ushort)(baseSizeInParagraphs + exeFile.MaxAlloc):
        freeSizeInParagraphs
);
```

Reply:
The reason I didn't make it read-only was because this function needs to change those values once it has calculated the largest free block. I can change the logic to do that, though. It will just require another temporary variable.